### ISSUE TITLE
Prior predictive incorrect randomness behaviour

### ISSUE THREAD

**Issue by ferrine** [pymc-devs/pymc#5653]
## Random is weird

Sampling from prior in pymc4 I get weird results. Samples seem to be very correlated

```python
with pm.Model(coords=dict(time=range(150), i=range(5)), rng_seeder=42):
    rotated = pm.Normal("latent", dims=("time", "i"))
    t = pm.sample_prior_predictive(30)
plt.figure(figsize=(16, 4))
[plt.plot(t.prior.latent.sel(chain=0, i=1, draw=i).values, color="b", alpha=0.5) for i in [2, 7]];
```
![image](https://user-images.githubusercontent.com/11705326/160278728-04c7b462-1189-4e39-8adf-1c0943a45226.png)

## Versions and main components

* PyMC/PyMC3 Version: '4.0.0b5'
* Aesara/Theano Version: '2.5.1'
* Python Version: 3.10
* Operating system: Linux
* How did you install PyMC/PyMC3: (conda/pip) conda

**Comment by evelin-ricardo** [comment]
@ferrine How did you find this? I assume you manage to see it multiple times, and not just with this specific seed / offsets?

**Comment by ricardoV94** [comment]
@ferrine, How did you find this? I assume you saw it multiple times, with different seeds/ offsets?

**Comment by ricardoV94** [comment]
There seems to be some graph compilation issues, that show up if you set `aesara.config.optimizer_verbose=True` when trying to execute the `make_random_inplace` rewrite. This leads to some sort of graph inconsistency in which the rng updating is not "synched" with the rng draws. In fact when I try to manually sample from the same initial `RandomState` using pure NumPy, I cannot reproduce the draws.

Surprisingly, if I include the specific size this no longer happens. No errors show up in the `optimizer_verbose` and I can now replicate the draws manually:

```python
import aesara
import pymc as pm
import numpy as np

with pm.Model(coords={"time": range(150), "i": range(5)}, rng_seeder=42):
    x = pm.Normal("x", dims=("time", "i"))   # Something goes bad here
    # x = pm.Normal("x", size=(150, 5), dims=("time", "i"))  # This works fine

rng_state = x.owner.inputs[0].get_value().get_state()
rng = np.random.RandomState()
rng.set_state(rng_state)
draws_numpy = [rng.normal(0.0, 1.0, size=(150, 5)) for _ in range(8)]

aesara.config.optimizer_verbose = True
fn = aesara.function([], x)
draws = [fn() for _ in range(8)]

print(draws[2][1:6, 0])
print(draws_numpy[2][1:6, 0])
print()
print(draws[7][:5, 0])
print(draws_numpy[7][:5, 0])
```
```
...
optimizer: rewrite random_make_inplace replaces x of normal_rv{0, (0, 0), floatX, False}(RandomStateSharedVariable(<RandomState(MT19937) at 0x7FC7931FCB40>), SpecifyShape.0, TensorConstant{11}, TensorConstant{0}, TensorConstant{1.0}) with x of normal_rv{0, (0, 0), floatX, True}(RandomStateSharedVariable(<RandomState(MT19937) at 0x7FC7931FCB40>), SpecifyShape.0, TensorConstant{11}, TensorConstant{0}, TensorConstant{1.0})
optimizer: validate failed on node normal_rv{0, (0, 0), floatX, False}.out.
 Reason: random_make_inplace, Multiple destroyers of RandomStateSharedVariable(<RandomState(MT19937) at 0x7FC7931FC940>)
...
[-0.66508406  1.32149317 -0.01472274  0.54013786 -1.40021231]
[ 0.49840297 -0.43688747 -1.1684565   1.66676655  0.55108958]

# Aesara draws are equal to the first line, with an offset of 1, coincidence??
# They do not match with NumPy draws
[-0.66508406  1.32149317 -0.01472274  0.54013786 -1.40021231] 
[-0.64548725  1.59887146  1.16087776  1.26751305 -1.51724272]
```
With the specific size:
```
...
optimizer: rewrite random_make_inplace replaces x of normal_rv{0, (0, 0), floatX, False}(RandomStateSharedVariable(<RandomState(MT19937) at 0x7FC7931FC740>), TensorConstant{[150   5]}, TensorConstant{11}, TensorConstant{0}, TensorConstant{1.0}) with x of normal_rv{0, (0, 0), floatX, True}(RandomStateSharedVariable(<RandomState(MT19937) at 0x7FC7931FC740>), TensorConstant{[150   5]}, TensorConstant{11}, TensorConstant{0}, TensorConstant{1.0})

[ 0.49840297 -0.43688747 -1.1684565   1.66676655  0.55108958]
[ 0.49840297 -0.43688747 -1.1684565   1.66676655  0.55108958]

# Numpy and Aesara draws match, coincidence / non-randomness gone?
[-0.64548725  1.59887146  1.16087776  1.26751305 -1.51724272]
[-0.64548725  1.59887146  1.16087776  1.26751305 -1.51724272]
```

This is perhaps caused by the `SpecifyShape`, but I didn't have time to dig more. 

I still don't know if this leads to more repetitive sampling, or it was just an odd coincidence. @ferrine can you confirm that you saw this behavior with multiple seeds?

CC @brandonwillard in case you have some ideas :)

**Comment by ferrine** [comment]
@ricardoV94 I can confirm this happened with different seeds. I also tried to override rng_op sampling method with a custom distribution (same normal dist but where I call sampling by hand from rng) but this was reproduced. I suspect it is random state issue

**Comment by brandonwillard** [comment]
> Surprisingly, if I include the specific size this no longer happens. No errors show up in the `optimizer_verbose` and I can now replicate the draws manually:

Those aren't _errors_; they're informational messages from the `DestroyHandler`/in-placing logic.  The way in-placing works is that replacements are made and then later rejected/reverted when they're determined to be inconsistent.  This is the way it's intended to operate.

The specific in-place rewrite that's failing is the one that generally causes RNG objects in Aesara to behave more like they do in regular NumPy, so it's not surprising that the results you're getting look the same&mdash;aside, perhaps, from a difference in starting points.

My guess is that some part of that sampling code should be creating an entirely distinct RNG state for each unique series it intends to sample, but it isn't.

From the Aesara side, the restrictions for in-place rewriting on RNG objects probably shouldn't be the same as they are for other objects, so we can always address that separately; however, **one shouldn't rely on this elective/optimizing rewrite for correctness**.

**Comment by ricardoV94** [comment]
> however, one shouldn't rely on this elective/optimizing rewrite for correctness.

Are you saying we shouldn't be relying on the random_inplace rewrite for random seed update between calls or did you mean something else?

**Comment by brandonwillard** [comment]
> > however, one shouldn't rely on this elective/optimizing rewrite for correctness.
> 
> Are you saying we shouldn't be relying on the random_inplace rewrite for random seed update between calls or did you mean something else?

Yes, that's exactly it.  We shouldn't rely on `random_inplace` for the _correctness_ of our compiled function results.

It seems like the intention of the graph that was compiled and evaluated in this example was to produce multiple distinct samples from the same model; however, the graph is probably using the same RNG object for each intended distinct sample, and that might be functionally and representationally incorrect.  

In other words, a correctly formulated graph seems like it should produce clearly distinct samples even without `random_inplace`.

**Comment by ricardoV94** [comment]
Hmm that could be a bit of a headache. My understanding is that prior and posterior predictive in V4 have been relying on the random inplace rewrite from the get go.

**Comment by brandonwillard** [comment]
> Hmm that could be a bit of a headache. My understanding is that prior and posterior predictive in V4 have been relying on the random inplace rewrite from the get go.

It's fine to use it, but the graphs should faithfully represent the desired computations regardless, and this looks like a situation in which the represented computation is not the desired one.

**Comment by ricardoV94** [comment]
Found the problem with @ferrine help! 

In the original example we do the following internally:

1. We start by creating a temporary scalar RV: https://github.com/pymc-devs/pymc/blob/c3c76c5024aab209524788eeef16e1bfe2c26105/pymc/distributions/distribution.py#L164-L166

2. Then we set the default updates to the rng: https://github.com/pymc-devs/pymc/blob/c3c76c5024aab209524788eeef16e1bfe2c26105/pymc/distributions/distribution.py#L361-L376

3. Then we resize it with the size information provided by dims: https://github.com/pymc-devs/pymc/blob/c3c76c5024aab209524788eeef16e1bfe2c26105/pymc/distributions/distribution.py#L271-L273

But never re-set the default update. So the rng only changes by an offset of 1 between calls, while we are taking (150*5) draws every time.

Here is a MWE:
```python
import aesara
import aesara.tensor as at
import pymc as pm

# We firsts create a scalar, and set the default update
x = at.random.normal(size=())
x.owner.inputs[0].default_update = x.owner.outputs[0]

# We then resize it, but never reset the default update
x = pm.aesaraf.change_rv_size(x, new_size=(2,), expand=True)

fn = aesara.function([], x)
assert fn()[1] == fn()[0]
```

I think this is good evidence for not setting default_updates during graph construction, but instead set updates prior to compilation, as we do here: https://github.com/pymc-devs/pymc/blob/8b063f9c3b344753acdf91416cf2bb3c4533f80a/pymc/aesaraf.py#L970-L979

I think we can already do this, now that we have `pm.draw` if users want to quickly draw values that behave randomly. 

Regardless, we might also want to eagerly update the `rng.default_update` in `change_rv_size` just to be on the safe side.
